src/hubextensions.ts

```diff
@@ -30,26 +31,74 @@ export function __PRIVATE__wrapStartTransactionWithProfiling(startTransaction: S
     const profile_id = uuid4();
     const uniqueTransactionName = `${transactionContext.name} ${profile_id}`;
 
-    // profilesSampleRate is multiplied with tracesSampleRate to get the final sampling rate.
+    // profilesSampleRate is multiplied with tracesSampleRate to get the final sampling rate. We dont perform
+    // the actual multiplication to get the final rate, but we discard the profile if the transaction was sampled,
+    // so anything after this block from here is based on the transaction sampling.
     if (!transaction.sampled) {
       return transaction;
     }
 
-    // @ts-expect-error profilesSampleRate is not part of the options type yet
-    const profilesSampleRate = this.getClient()?.getOptions?.()?.profilesSampleRate;
-    if (profilesSampleRate === undefined) {
+    const client = this.getClient();
+    if (!client) {
+      if (isDebugBuild()) {
+        logger.log('[Profiling] Profiling disabled, no client found.');
+      }
+      return transaction;
+    }
+    const options = client.getOptions();
+    if (!options) {
+      if (isDebugBuild()) {
+        logger.log('[Profiling] Profiling disabled, no options found.');
+      }
+      return transaction;
+    }
+
+    // @ts-expect-error sampler is not yer part of the sdk options
+    const profilesSampler = options.profilesSampler;
+    // @ts-expect-error sampler is not yer part of the sdk options
+    let profilesSampleRate: number | undefined = options.profilesSampleRate;
+
+    // Prefer sampler to sample rate if both are provided.
+    if (typeof profilesSampler === 'function') {
+      profilesSampleRate = profilesSampler(customSamplingContext);
+    } else if (customSamplingContext && customSamplingContext['parentSampled'] !== undefined) {
+      profilesSampleRate = customSamplingContext['parentSampled'];
+    }
+
+    // Since this is coming from the user (or from a function provided by the user), who knows what we might get. (The
+    // only valid values are booleans or numbers between 0 and 1.)
+    if (!isValidSampleRate(profilesSampleRate)) {
+      if (isDebugBuild()) {
+        logger.warn('[Profiling] Discarding profile because of invalid sample rate.');
+      }
+      return transaction;
+    }
+
+    // if the function returned 0 (or false), or if `profileSampleRate` is 0, it's a sign the profile should be dropped
+    if (!profilesSampleRate) {
```

Comment:
Does this case need to exist? It gets handled by the `if (!sampled)` check below right?

Reply:
I think so, at least the tracing SDK seems to do this + the reasons are subtly different. The first check checks if profilesSampleRate was 0 or inherited 0 and second check checks if we are discarding as a consequence of the dice roll.